cnf-certification-test/observability/pdb/pdb.go

```diff
@@ -14,13 +29,52 @@ func CheckPDBIsValid(pdb *policyv1.PodDisruptionBudget, replicas *int32) (bool,
 		replicaCount = 1 // default value
 	}
 
-	if pdb.Spec.MinAvailable != nil && pdb.Spec.MinAvailable.IntValue() == 0 {
-		return false, fmt.Errorf("field .spec.minAvailable cannot be zero. Currently set to: %d", pdb.Spec.MinAvailable.IntValue())
+	var minAvailableValue int
+	var maxUnavailableValue int
+
+	if pdb.Spec.MinAvailable != nil {
+		var err error
+		minAvailableValue, err = intOrStringToValue(pdb.Spec.MinAvailable, replicaCount)
+		if err != nil {
+			return false, err
+		}
+
+		// Tests for the minAvailable spec
+		if minAvailableValue == 0 {
+			return false, fmt.Errorf("field .spec.minAvailable cannot be zero. Currently set to: %d", minAvailableValue)
+		}
+
+		if minAvailableValue > int(replicaCount) {
+			return false, fmt.Errorf("minAvailable cannot be greater than replicas. Currently set to: %d", minAvailableValue)
+		}
 	}
 
-	if pdb.Spec.MaxUnavailable != nil && pdb.Spec.MaxUnavailable.IntValue() >= int(replicaCount) {
-		return false, fmt.Errorf("field .spec.maxUnavailable cannot be greater than or equal to the number of pods in the replica. Currently set to: %d", pdb.Spec.MaxUnavailable.IntValue())
+	if pdb.Spec.MaxUnavailable != nil {
+		var err error
+		maxUnavailableValue, err = intOrStringToValue(pdb.Spec.MaxUnavailable, replicaCount)
+		if err != nil {
+			return false, err
+		}
+
+		// Tests for the maxUnavailable spec
+		if maxUnavailableValue >= int(replicaCount) {
+			return false, fmt.Errorf("field .spec.maxUnavailable cannot be greater than or equal to the number of pods in the replica. Currently set to: %d", maxUnavailableValue)
+		}
 	}
 
 	return true, nil
 }
+
+func intOrStringToValue(intOrStr *intstr.IntOrString, replicas int32) (int, error) {
+	switch intOrStr.Type {
+	case intstr.Int:
+		return intOrStr.IntValue(), nil
+	case intstr.String:
+		v, err := percentageToFloat(intOrStr.StrVal)
+		if err != nil {
+			return 0, fmt.Errorf("invalid value %q: %v", intOrStr.StrVal, err)
+		}
+		return int(v * float64(replicas)), nil
```

Comment:
use int(math.Round(v * float64(replicas))), see: https://kubernetes.io/docs/tasks/run-application/configure-pdb/
When you specify the value as a percentage, it may not map to an exact number of Pods. For example, if you have 7 Pods and you set minAvailable to "50%", it's not immediately obvious whether that means 3 Pods or 4 Pods must be available. Kubernetes rounds up to the nearest integer, so in this case, 4 Pods must be available.

Reply:
Right yup I saw that document earlier.  Good idea to use the `math.Round` I'll fix.